store/etcd/get.go

```diff
@@ -13,113 +13,63 @@ import (
 	}
-	entry.Versions = v.Versions
-
-	resp, err = s.Client.KV.Get(ctx, s.signaturesPath(path))
+	// running all the requests within a single transaction so only one network round trip is performed.
+	resp, err := s.Client.KV.Txn(ctx).Then(ops...).Commit()
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to fetch signature of the entry: %s", path)
-	}
-	if resp.Count == 0 {
-		s.Logger.Debug().Str("path", path).Msg("cannot find ruleset signature")
-		return nil, store.ErrNotFound
+		return nil, errors.Wrapf(err, "failed to fetch ruleset: %s", path)
 	}
 
-	var sig pb.Signature
-	err = proto.Unmarshal(resp.Kvs[0].Value, &sig)
-	if err != nil {
-		return nil, errors.Wrapf(err, "failed to unmarshal signature")
+	if len(resp.Responses) != 3 {
+		return nil, store.ErrRulesetNotFound
 	}
-	entry.Signature = signatureFromProtobuf(&sig)
-
-	return entry, nil
-}
 
-// Latest returns the latest version of the ruleset entry which corresponds to the given path.
-// It returns store.ErrNotFound if the path doesn't exist or if it's not a ruleset.
-func (s *RulesetService) Latest(ctx context.Context, path string) (*store.RulesetEntry, error) {
-	if path == "" {
-		return nil, store.ErrNotFound
+	// decode signature
+	var sig pb.Signature
+	gresp := resp.Responses[0].GetResponseRange()
+	if gresp.Count == 0 {
+		return nil, store.ErrRulesetNotFound
 	}
-
-	resp, err := s.Client.KV.Get(ctx, s.entriesPath(path, "")+versionSeparator, clientv3.WithLastKey()...)
+	err = proto.Unmarshal(gresp.Kvs[0].Value, &sig)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to fetch the entry: %s", path)
-	}
-
-	// Count will be 0 if the path doesn't exist or if it's not a ruleset.
-	if resp.Count == 0 {
-		return nil, store.ErrNotFound
+		return nil, errors.Wrap(err, "failed to unmarshal signature")
 	}
 
-	var entry pb.RulesetEntry
-	err = proto.Unmarshal(resp.Kvs[0].Value, &entry)
+	// decode versions
+	var versions pb.Versions
+	err = proto.Unmarshal(resp.Responses[1].GetResponseRange().Kvs[0].Value, &versions)
 	if err != nil {
-		s.Logger.Debug().Err(err).Bytes("entry", resp.Kvs[0].Value).Msg("latest: unmarshalling failed")
-		return nil, errors.Wrap(err, "failed to unmarshal entry")
+		return nil, errors.Wrap(err, "failed to unmarshal versions")
 	}
 
-	return &store.RulesetEntry{
-		Path:    entry.Path,
-		Version: entry.Version,
-		Ruleset: rulesetFromProtobuf(entry.Ruleset),
-	}, nil
-}
-
-// OneByVersion returns the ruleset entry which corresponds to the given path at the given version.
-// It returns store.ErrNotFound if the path doesn't exist or if it's not a ruleset.
-func (s *RulesetService) OneByVersion(ctx context.Context, path, version string) (*store.RulesetEntry, error) {
-	if path == "" {
-		return nil, store.ErrNotFound
-	}
-
-	resp, err := s.Client.KV.Get(ctx, s.entriesPath(path, version))
+	// decode ruleset
+	var ruleset pb.Ruleset
+	err = proto.Unmarshal(resp.Responses[2].GetResponseRange().Kvs[0].Value, &ruleset)
```

Comment:
If we just have a signature without rulesets, it will panic here, the same above for the versions.
I understood that you make the verification once for the whole entities (L.43) but it is not relevant if we just have a signature without a rulesets "added" to it.